# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/DriftCompensation.h
# Commit: b8855a45d798
# Full Hash: b8855a45d798aac9bdf82cd328eb17375c5510e5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-04-08 19:00:29
# Description:
#   Bug 1542685 - Avoid integer overflows by multiplying doubles. r=padenot
#   
#   SaferMultDiv(time, audioScale, videoScale) could easily result in overflow
#   because all three args are roughly equal, and SaferMultDiv would always do the
#   multiplication first. The worst-case is then multiplying an int64_t to another
# ==============================================================================

diff -r b306079acdcb -r b8855a45d798 dom/media/DriftCompensation.h
--- a/dom/media/DriftCompensation.h	Thu Apr 04 12:52:22 2019 +0000
+++ b/dom/media/DriftCompensation.h	Mon Apr 08 14:27:13 2019 +0000
@@ -103,24 +103,25 @@
       return aTime;
     }
 
-    int64_t videoScaleUs = (aNow - mAudioStartTime).ToMicroseconds();
-    int64_t audioScaleUs = FramesToUsecs(samples, mAudioRate).value();
-    int64_t videoDurationUs = (aTime - mAudioStartTime).ToMicroseconds();
-
-    if (videoScaleUs == 0) {
-      videoScaleUs = audioScaleUs;
+    if (aNow == mAudioStartTime) {
+      LOG(LogLevel::Warning,
+          "DriftCompensator %p video scale 0, assuming no drift", this);
+      return aTime;
     }
 
+    double videoScaleUs = (aNow - mAudioStartTime).ToMicroseconds();
+    double audioScaleUs = FramesToUsecs(samples, mAudioRate).value();
+    double videoDurationUs = (aTime - mAudioStartTime).ToMicroseconds();
+
     TimeStamp reclocked =
-        mAudioStartTime +
-        TimeDuration::FromMicroseconds(
-            SaferMultDiv(videoDurationUs, audioScaleUs, videoScaleUs).value());
+        mAudioStartTime + TimeDuration::FromMicroseconds(
+                              videoDurationUs * audioScaleUs / videoScaleUs);
 
     LOG(LogLevel::Debug,
         "DriftCompensator %p GetVideoTime, v-now: %.3fs, a-now: %.3fs; %.3fs "
         "-> %.3fs (d %.3fms)",
         this, (aNow - mAudioStartTime).ToSeconds(),
-        static_cast<double>(audioScaleUs) / 1000000.0,
+        TimeDuration::FromMicroseconds(audioScaleUs).ToSeconds(),
         (aTime - mAudioStartTime).ToSeconds(),
         (reclocked - mAudioStartTime).ToSeconds(),
         (reclocked - aTime).ToMilliseconds());
